src/lib/core/data-table/data-table.ts

```diff
@@ -77,17 +94,34 @@ export class CdkTable<T> implements CollectionViewer {
   viewChange =
       new BehaviorSubject<{start: number, end: number}>({start: 0, end: Number.MAX_VALUE});
 
-  /** Stream that emits when a row def has a change to its array of columns to render. */
-  _columnsChange = new Observable<void>();
-
   /**
-   * Map of all the user's defined columns identified by name.
-   * Contains the header and data-cell templates.
+   * Provides a stream containing the latest data array to render. Influenced by the table's
+   * stream of view window (what rows are currently on screen).
    */
-  private _columnDefinitionsByName = new Map<string,  CdkColumnDef>();
-
-  /** Differ used to find the changes in the data provided by the data source. */
-  private _dataDiffer: IterableDiffer<T> = null;
+  @Input()
+  get dataSource(): DataSource<T> { return this._dataSource; }
+  set dataSource(dataSource: DataSource<T>) {
+    if (this._dataSource == dataSource) { return; }
+    this._data = [];
```

Comment:
Can you move most of this setter out into a separate function like `_updateDataSource` or `_switchDataSource`?